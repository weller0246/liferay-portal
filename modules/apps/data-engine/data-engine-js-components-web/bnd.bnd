Bundle-Name: Liferay Data Engine JS Components Web
Bundle-SymbolicName: com.liferay.data.engine.js.components.web
Bundle-Version: 1.0.18
Web-ContextPath: /data-engine-js-components-web