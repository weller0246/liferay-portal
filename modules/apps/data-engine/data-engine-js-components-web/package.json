{
	"dependencies": {
		"@clayui/button": "3.32.0",
		"@clayui/data-provider": "3.35.3",
		"@clayui/drop-down": "3.35.3",
		"@clayui/icon": "3.32.0",
		"@clayui/layout": "3.32.0",
		"@clayui/loading-indicator": "3.32.0",
		"@clayui/pagination": "3.35.3",
		"@clayui/tabs": "3.35.3",
		"@clayui/tooltip": "3.35.3",
		"classnames": "2.3.1",
		"dom-align": "1.10.4",
		"frontend-js-web": "*",
		"incremental-dom-string": "0.0.2",
		"metal-component": "2.16.8",
		"metal-soy": "2.16.9",
		"moment": "^2.22.2",
		"react": "16.12.0",
		"react-dnd": "11.1.1",
		"react-dnd-html5-backend": "11.1.1",
		"react-dom": "16.12.0"
	},
	"jest": {
		"collectCoverageFrom": [
			"src/main/resources/META-INF/resources/js/**/*.es.js"
		],
		"modulePathIgnorePatterns": [
			"/__mock__/"
		],
		"setupFiles": [
			"<rootDir>/jest-setup.config.js"
		],
		"snapshotSerializers": [
			"metal-jest-serializer"
		],
		"testPathIgnorePatterns": [
			"<rootDir>/test/js/[^/]+\\.js"
		]
	},
	"main": "js/index.es.js",
	"name": "data-engine-js-components-web",
	"private": true,
	"scripts": {
		"build": "liferay-npm-scripts build",
		"checkFormat": "liferay-npm-scripts check",
		"format": "liferay-npm-scripts fix",
		"test": "metalsoy -s \"test/**/*.soy\" -d \"test\" && liferay-npm-scripts test"
	},
	"version": "1.0.18"
}
