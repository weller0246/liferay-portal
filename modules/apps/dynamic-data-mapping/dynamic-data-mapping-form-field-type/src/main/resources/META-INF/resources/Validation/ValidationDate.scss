.ddm-form-field-type__validation-date {
	width: 33%;
}
