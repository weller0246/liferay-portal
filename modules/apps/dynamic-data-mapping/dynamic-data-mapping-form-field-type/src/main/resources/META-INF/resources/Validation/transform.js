/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

const getValidationFromExpression = (validations, validation) => {
	return function transformValidationFromExpression(expression) {
		let mutValidation;

		if (!expression && validation) {
			expression = validation.expression;
		}

		if (expression) {
			mutValidation = validations.find(
				(validation) => validation.name === expression.name
			);
		}

		return mutValidation;
	};
};

const transformValidations = (validations, dataType) => {
	return validations[normalizeDataType(dataType)].map((validation) => {
		return {
			...validation,
			checked: false,
			value: validation.name,
		};
	});
};

const getValidation = (validations, transformValidationFromExpression) => {
	return function transformValue(value) {
		const {errorMessage = {}, expression = {}, parameter = {}} = value;
		let parameterMessage = '';
		let selectedValidation = transformValidationFromExpression(expression);
		const enableValidation = !!expression.value;

		if (selectedValidation) {
			parameterMessage = selectedValidation.parameterMessage;
		}
		else {
			selectedValidation = validations[0];
		}

		return {
			enableValidation,
			errorMessage,
			expression,
			parameter,
			parameterMessage,
			selectedValidation,
		};
	};
};

export const normalizeDataType = (initialDataType) => {
	return initialDataType === 'double' || initialDataType === 'integer'
		? 'numeric'
		: initialDataType;
};

export const getLocalizedValue = ({defaultLanguageId, editingLanguageId}) => (
	value
) => value[editingLanguageId] || value[defaultLanguageId];

export const getSelectedValidation = (validations) => {
	return function transformSelectedValidation(value) {
		if (Array.isArray(value)) {
			value = value[0];
		}

		let selectedValidation = validations.find(({name}) => name === value);

		if (!selectedValidation) {
			selectedValidation = validations[0];
		}

		return selectedValidation;
	};
};

export const transformData = ({
	defaultLanguageId,
	editingLanguageId,
	initialDataType,
	validation,
	validations: initialValidations,
	value,
}) => {
	const dataType = validation?.dataType ?? initialDataType;
	const validations = transformValidations(initialValidations, dataType);
	const parsedValidation = getValidation(
		validations,
		getValidationFromExpression(validations, validation)
	)(value);
	const localizationMode = editingLanguageId !== defaultLanguageId;

	return {
		...parsedValidation,
		dataType,
		localizationMode,
		validations,
	};
};
