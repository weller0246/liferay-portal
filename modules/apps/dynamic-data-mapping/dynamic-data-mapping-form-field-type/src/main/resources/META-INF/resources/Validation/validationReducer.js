/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import {StringUtils} from 'data-engine-js-components-web';

export const EVENT_TYPES = {
	CHANGE_ERROR_MESSAGE: 'changeErrorMessage',
	ENABLE_VALIDATION: 'enableValidation',
	SELECTED_VALIDATION: 'changeSelectedValidation',
	SET_PARAMETER: 'setParameter',
};

const reducer = ({editingLanguageId, fieldName, onChange}) => (
	state,
	action
) => {
	const onSave = (state) => {

		// Expression is mounted in the frontend, so we need to update
		// the expression every time we update the state

		let expression = {};

		if (state.enableValidation) {
			expression = {
				name: state.selectedValidation.name,
				value: StringUtils.subWords(state.selectedValidation.template, {
					name: fieldName,
				}),
			};
		}

		// Sends these values up using onChange, this way the data is saved

		const {enableValidation, errorMessage, parameter} = state;

		onChange({
			enableValidation,
			errorMessage,
			expression,
			parameter,
		});

		return {
			...state,
			expression,
		};
	};

	switch (action.type) {
		case EVENT_TYPES.ENABLE_VALIDATION: {
			const {enableValidation} = action.payload;

			return onSave({
				...state,
				enableValidation,
			});
		}
		case EVENT_TYPES.CHANGE_ERROR_MESSAGE: {
			const {errorMessage} = action.payload;

			return onSave({
				...state,
				errorMessage: {
					...state.errorMessage,
					[editingLanguageId]: errorMessage,
				},
			});
		}
		case EVENT_TYPES.CHANGE_SELECTED_VALIDATION: {
			const {selectedValidation} = action.payload;

			return onSave({
				...state,
				selectedValidation,
			});
		}
		case EVENT_TYPES.SET_PARAMETER: {
			const {parameter} = action.payload;

			return onSave({
				...state,
				parameter: {
					...state.parameter,
					[editingLanguageId]: parameter,
				},
			});
		}
		default:
			return state;
	}
};

export default reducer;
