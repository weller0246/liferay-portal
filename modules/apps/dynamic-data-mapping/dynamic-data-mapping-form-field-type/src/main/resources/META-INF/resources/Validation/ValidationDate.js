/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import './ValidationDate.scss';

import {ClayInput} from '@clayui/form';
import React, {useState} from 'react';

import Select from '../Select/Select.es';
import {EVENT_TYPES} from './validationReducer';

const MAX_QUANTITY = 999;
const MIN_QUANTITY = 1;

const customDateLabel = Liferay.Language.get('custom-date');
const daysLabel = Liferay.Language.get('days');
const endsOnLabel = Liferay.Language.get('ends-on');
const minusLabel = Liferay.Language.get('minus');
const monthsLabel = Liferay.Language.get('months');
const plusLabel = Liferay.Language.get('plus');
const responseDateLabel = Liferay.Language.get('response-date');
const startsFromLabel = Liferay.Language.get('starts-from');
const yearsLabel = Liferay.Language.get('years');

const responseDateOption = {
	checked: false,
	label: responseDateLabel,
	name: 'responseDate',
	value: 'responseDate',
};

const operationsOptions = [
	{
		checked: false,
		label: minusLabel,
		name: 'minus',
		value: 'minus',
	},
	{
		checked: false,
		label: plusLabel,
		name: 'plus',
		value: 'plus',
	},
];

const unitOptions = [
	{
		checked: false,
		label: daysLabel,
		name: 'days',
		value: 'days',
	},
	{
		checked: false,
		label: monthsLabel,
		name: 'months',
		value: 'months',
	},
	{
		checked: false,
		label: yearsLabel,
		name: 'years',
		value: 'years',
	},
];

const getDateOptionsByType = (label, name) => ({
	label,
	name,
	options: [
		responseDateOption,
		{
			checked: false,
			label: customDateLabel,
			name: 'customDate',
			value: 'customDate',
		},
	],
});

const getOperation = (quantity) => {
	return quantity < 0 ? 'minus' : 'plus';
};

const getSelectedParameter = (
	localizedValue,
	selectedParameterName,
	attributeName
) => {
	if (localizedValue && typeof localizedValue === 'string') {
		try {
			localizedValue = JSON.parse(localizedValue);
		}
		catch (error) {}
	}

	return localizedValue?.[selectedParameterName]?.[attributeName];
};

const getSignedValue = (operation, value) => {
	let signedValue = value;

	if (
		(operation === 'minus' && value > 0) ||
		(operation === 'plus' && value < 0)
	) {
		signedValue = value * -1;
	}

	return signedValue;
};

const parameters = {
	dateRange: [
		getDateOptionsByType(startsFromLabel, 'startsFrom'),
		getDateOptionsByType(endsOnLabel, 'endsOn'),
	],
	futureDates: [getDateOptionsByType(startsFromLabel, 'startsFrom')],
	pastDates: [getDateOptionsByType(endsOnLabel, 'endsOn')],
};

const CustomDates = ({
	date,
	eventType,
	handleChangeParameters,
	localizationMode,
	operation,
	quantity,
	readOnly,
	setDate,
	setOperation,
	setQuantity,
	setUnit,
	unit,
	visible,
}) => {
	return (
		<>
			<Select
				label={Liferay.Language.get('date')}
				name={`selectedDate_${eventType}`}
				onChange={(event, value) => {
					setDate(value[0]);
					handleChangeParameters(value[0], eventType, 'date');
				}}
				options={[responseDateOption]}
				readOnly={readOnly || localizationMode}
				showEmptyOption={false}
				value={date}
				visible={visible}
			/>
			<div className="align-items-end d-flex position-relative">
				<div className="ddm-form-field-type__validation-date pr-2">
					<Select
						label={Liferay.Language.get('operation')}
						name={`selectedOperation_${eventType}`}
						onChange={(event, value) => {
							setOperation(value[0]);
							handleChangeParameters(
								value[0],
								eventType,
								'quantity'
							);
						}}
						options={operationsOptions}
						readOnly={readOnly || localizationMode}
						showEmptyOption={false}
						value={operation}
						visible={visible}
					/>
				</div>
				<div className="ddm-form-field-type__validation-date pr-2">
					<div className="form-group">
						<label htmlFor="validation_date_quantity">
							{Liferay.Language.get('quantity')}
						</label>
						<ClayInput
							className="ddm-field-text"
							disabled={readOnly}
							id="validation_date_quantity"
							max={MAX_QUANTITY}
							name={`inputedQuantity_${eventType}`}
							onBlur={(event) => {
								let {value: newValue} = event.target;

								if (newValue > MAX_QUANTITY) {
									newValue = MAX_QUANTITY;
								}
								else if (newValue < MIN_QUANTITY) {
									newValue = MIN_QUANTITY;
								}

								newValue =
									operation === 'minus'
										? (newValue * -1).toString()
										: newValue;

								setQuantity(newValue);
								handleChangeParameters(
									newValue,
									eventType,
									'quantity'
								);
							}}
							onChange={({target: {value}}) => {
								setQuantity(value);
							}}
							type="number"
							value={quantity === '' ? '' : Math.abs(quantity)}
						/>
					</div>
				</div>
				<div className="ddm-form-field-type__validation-date">
					<Select
						label={Liferay.Language.get('unit')}
						name={`selectedUnit_${eventType}`}
						onChange={(event, value) => {
							setUnit(value[0]);
							handleChangeParameters(value[0], eventType, 'unit');
						}}
						options={unitOptions}
						readOnly={readOnly || localizationMode}
						showEmptyOption={false}
						value={unit}
						visible={visible}
					/>
				</div>
			</div>
		</>
	);
};

const ValidationDate = ({
	dispatch,
	errorMessage,
	localizationMode,
	localizedValue,
	name,
	onBlur,
	parameter,
	readOnly,
	selectedValidation,
	transformSelectedValidation,
	validations,
	visible,
}) => {
	const initialStartQuantity = getSelectedParameter(
		localizedValue(parameter),
		'startsFrom',
		'quantity'
	);
	const initialEndQuantity = getSelectedParameter(
		localizedValue(parameter),
		'endsOn',
		'quantity'
	);
	const initialStartDate = getSelectedParameter(
		localizedValue(parameter),
		'startsFrom',
		'date'
	);
	const initialEndDate = getSelectedParameter(
		localizedValue(parameter),
		'endsOn',
		'date'
	);

	const initialStartType = getSelectedParameter(
		localizedValue(parameter),
		'startsFrom',
		'type'
	);
	const initialEndType = getSelectedParameter(
		localizedValue(parameter),
		'endsOn',
		'type'
	);
	const initialStartUnit = getSelectedParameter(
		localizedValue(parameter),
		'startsFrom',
		'unit'
	);
	const initialEndUnit = getSelectedParameter(
		localizedValue(parameter),
		'endsOn',
		'unit'
	);

	const selectedParameter = parameters[selectedValidation.name];
	const [startDate, setStartDate] = useState(initialStartDate);
	const [startOperation, setStartOperation] = useState(
		getOperation(initialStartQuantity)
	);
	const [startQuantity, setStartQuantity] = useState(initialStartQuantity);
	const [startsFrom, setStartsFrom] = useState(initialStartType);

	const [startUnit, setStartUnit] = useState(initialStartUnit);

	const [endDate, setEndDate] = useState(initialEndDate);
	const [endOperation, setEndOperation] = useState(
		getOperation(initialEndQuantity)
	);
	const [endQuantity, setEndQuantity] = useState(initialEndQuantity);
	const [endsOn, setEndsOn] = useState(initialEndType);
	const [endUnit, setEndUnit] = useState(initialEndUnit);

	const handleChangeParameters = (value, typeName, type) => {
		const parameter = {};

		let newValue = value;

		if (startsFrom) {
			parameter['startsFrom'] = {
				date: startDate,
				quantity: startQuantity,
				type: startsFrom,
				unit: startUnit,
			};
		}

		if (endsOn) {
			parameter['endsOn'] = {
				date: endDate,
				quantity: endQuantity,
				type: endsOn,
				unit: endUnit,
			};
		}

		let operation =
			typeName === 'startsFrom' ? startOperation : endOperation;

		if (type === 'quantity') {
			let quantity = value;

			if (isNaN(value)) {
				operation = value;
				quantity =
					typeName === 'startsFrom' ? startQuantity : endQuantity;
			}

			newValue = getSignedValue(operation, quantity);
		}

		parameter[typeName] = {
			...parameter[typeName],
			[type]: newValue,
		};

		dispatch({payload: {parameter}, type: EVENT_TYPES.SET_PARAMETER});
	};

	const errorMessageName = name + '_errorMessage';
	const getDateTypeValue = (name) => {
		return name === 'startsFrom' ? startsFrom : endsOn;
	};

	return (
		<>
			<Select
				disableEmptyOption
				label={Liferay.Language.get('accepted-date')}
				name="selectedValidation"
				onChange={(event, value) => {
					setStartsFrom(initialStartType);
					setEndsOn(initialEndType);
					dispatch({
						payload: {
							selectedValidation: transformSelectedValidation(
								value
							),
						},
						type: EVENT_TYPES.CHANGE_SELECTED_VALIDATION,
					});
				}}
				options={validations}
				placeholder={Liferay.Language.get('choose-an-option')}
				readOnly={readOnly || localizationMode}
				showEmptyOption={false}
				value={[selectedValidation.name]}
				visible={visible}
			/>
			{selectedParameter.map((element, index) => {
				let label = '';
				const startSection = element.name === 'startsFrom';

				if (selectedParameter.length > 1) {
					label = startSection
						? Liferay.Language.get('start-date')
						: Liferay.Language.get('end-date');
				}

				return (
					<>
						{label !== '' && (
							<>
								<label>{label.toUpperCase()}</label>
								<div className="separator" />
							</>
						)}

						<Select
							disableEmptyOption
							key={`selectedParameter_${index}`}
							label={element.label}
							name="selectedParameter"
							onChange={(event, value) => {
								if (startSection) {
									setStartsFrom(value[0]);
									setStartOperation(
										getOperation(initialStartQuantity)
									);
									setStartQuantity(initialStartQuantity);
									setStartDate(initialStartDate);
								}
								else {
									setEndsOn(value[0]);
									setEndOperation(
										getOperation(initialEndQuantity)
									);
									setEndQuantity(initialEndQuantity);
									setEndDate(initialEndDate);
								}
								handleChangeParameters(
									value[0],
									element.name,
									'type'
								);
							}}
							options={selectedParameter[index].options}
							placeholder={Liferay.Language.get(
								'choose-an-option'
							)}
							readOnly={localizationMode || readOnly}
							showEmptyOption={false}
							value={startSection ? startsFrom : endsOn}
							visible={visible}
						/>

						{getDateTypeValue(element.name) === 'customDate' && (
							<CustomDates
								date={startSection ? startDate : endDate}
								endDate={endDate}
								eventType={element.name}
								handleChangeParameters={handleChangeParameters}
								operation={
									startSection ? startOperation : endOperation
								}
								quantity={
									startSection ? startQuantity : endQuantity
								}
								readOnly={localizationMode || readOnly}
								setDate={
									startSection ? setStartDate : setEndDate
								}
								setOperation={
									startSection
										? setStartOperation
										: setEndOperation
								}
								setQuantity={
									startSection
										? setStartQuantity
										: setEndQuantity
								}
								setUnit={
									startSection ? setStartUnit : setEndUnit
								}
								unit={startSection ? startUnit : endUnit}
								visible={visible}
							/>
						)}
					</>
				);
			})}
			<label htmlFor={errorMessageName}>
				{Liferay.Language.get('error-message')}
			</label>
			<ClayInput
				id={errorMessageName}
				name={errorMessageName}
				onBlur={onBlur}
				onChange={(event) => {
					dispatch({
						payload: {
							errorMessage: event.target.value,
						},
						type: EVENT_TYPES.CHANGE_ERROR_MESSAGE,
					});
				}}
				placeholder={Liferay.Language.get('this-field-is-invalid')}
				type="text"
				value={localizedValue(errorMessage)}
			/>
		</>
	);
};

export default ValidationDate;
