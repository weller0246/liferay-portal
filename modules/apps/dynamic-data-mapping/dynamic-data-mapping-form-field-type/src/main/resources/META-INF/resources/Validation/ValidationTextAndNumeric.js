/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import React from 'react';

import Numeric from '../Numeric/Numeric';
import Select from '../Select/Select.es';
import Text from '../Text/Text.es';
import {EVENT_TYPES} from './validationReducer';

const ValidationTextAndNumeric = ({
	dataType,
	dispatch,
	errorMessage,
	localizationMode,
	localizedValue,
	name,
	onBlur,
	parameter,
	readOnly,
	selectedValidation,
	transformSelectedValidation,
	validations,
	visible,
}) => {
	const DynamicComponent =
		selectedValidation &&
		selectedValidation.parameterMessage &&
		dataType === 'string'
			? Text
			: Numeric;

	return (
		<>
			<Select
				disableEmptyOption
				label={Liferay.Language.get('accept-if-input')}
				name="selectedValidation"
				onChange={(event, value) => {
					dispatch({
						payload: {
							selectedValidation: transformSelectedValidation(
								value
							),
						},
						type: EVENT_TYPES.CHANGE_SELECTED_VALIDATION,
					});
				}}
				options={validations}
				placeholder={Liferay.Language.get('choose-an-option')}
				readOnly={readOnly || localizationMode}
				showEmptyOption={false}
				value={[selectedValidation.name]}
				visible={visible}
			/>

			{selectedValidation.parameterMessage && (
				<DynamicComponent
					dataType={dataType}
					label={Liferay.Language.get('value')}
					name={`${name}_parameter`}
					onChange={(event) => {
						dispatch({
							payload: {
								parameter: event.target.value,
							},
							type: EVENT_TYPES.SET_PARAMETER,
						});
					}}
					placeholder={selectedValidation.parameterMessage}
					readOnly={readOnly}
					required={false}
					value={localizedValue(parameter)}
					visible={visible}
				/>
			)}

			<Text
				label={Liferay.Language.get('error-message')}
				name={`${name}_errorMessage`}
				onBlur={onBlur}
				onChange={(event) => {
					dispatch({
						payload: {
							errorMessage: event.target.value,
						},
						type: EVENT_TYPES.CHANGE_ERROR_MESSAGE,
					});
				}}
				placeholder={Liferay.Language.get('this-field-is-invalid')}
				readOnly={readOnly}
				required={false}
				value={localizedValue(errorMessage)}
				visible={visible}
			/>
		</>
	);
};

export default ValidationTextAndNumeric;
