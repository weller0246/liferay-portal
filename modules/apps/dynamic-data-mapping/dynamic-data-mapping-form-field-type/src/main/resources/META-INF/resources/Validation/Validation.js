/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import ClayForm from '@clayui/form';
import {useFormState} from 'data-engine-js-components-web';
import React, {useEffect, useReducer} from 'react';

import Checkbox from '../Checkbox/Checkbox';
import ValidationDate from './ValidationDate';
import ValidationTextAndNumeric from './ValidationTextAndNumeric';
import {
	getLocalizedValue,
	getSelectedValidation,
	normalizeDataType,
	transformData,
} from './transform';
import reducer, {EVENT_TYPES} from './validationReducer';

const COMPONENTS = {
	date: ValidationDate,
	numeric: ValidationTextAndNumeric,
	string: ValidationTextAndNumeric,
};

const Validation = ({
	dataType,
	defaultLanguageId,
	editingLanguageId,
	enableValidation: initialEnableValidation,
	errorMessage: initialErrorMessage,
	label,
	localizationMode,
	name,
	onBlur,
	onChange,
	parameter: initialParameter,
	readOnly,
	selectedValidation: initialSelectedValidation,
	validation,
	validations,
	visible,
}) => {
	const {focusedField} = useFormState();

	const [
		{enableValidation, errorMessage, parameter, selectedValidation},
		dispatch,
	] = useReducer(
		reducer({
			editingLanguageId,
			fieldName: validation?.fieldName ?? focusedField?.fieldName,
			onChange,
		}),
		{
			enableValidation: initialEnableValidation,
			errorMessage: initialErrorMessage,
			parameter: initialParameter,
			selectedValidation: initialSelectedValidation,
		}
	);

	const ValidationComponent = COMPONENTS[normalizeDataType(dataType)];

	const transformSelectedValidation = getSelectedValidation(validations);

	const localizedValue = getLocalizedValue({
		defaultLanguageId,
		editingLanguageId,
	});

	useEffect(() => {
		if (readOnly || !visible) {
			dispatch({
				payload: {enableValidation: false},
				type: EVENT_TYPES.ENABLE_VALIDATION,
			});
		}
	}, [readOnly, visible]);

	return (
		<ClayForm.Group className="lfr-ddm-form-field-validation">
			<Checkbox
				label={label}
				name="enableValidation"
				onChange={({target: {value: enableValidation}}) => {
					dispatch({
						payload: {
							enableValidation,
							fieldName: validation?.fieldName,
						},
						type: EVENT_TYPES.ENABLE_VALIDATION,
					});
				}}
				readOnly={readOnly}
				showAsSwitcher
				value={enableValidation}
				visible={visible}
			/>

			{enableValidation && (
				<ValidationComponent
					dataType={dataType}
					dispatch={dispatch}
					errorMessage={errorMessage}
					localizationMode={localizationMode}
					localizedValue={localizedValue}
					name={name}
					onBlur={onBlur}
					parameter={parameter}
					readOnly={readOnly}
					selectedValidation={selectedValidation}
					transformSelectedValidation={transformSelectedValidation}
					validations={validations}
					visible={visible}
				/>
			)}
		</ClayForm.Group>
	);
};

const ValidationWrapper = ({
	dataType: initialDataType,
	defaultLanguageId,
	editingLanguageId,
	label,
	name,
	onBlur,
	onChange,
	readOnly,
	validation,
	value,
	visible,
}) => {
	const {validations} = useFormState();
	const data = transformData({
		defaultLanguageId,
		editingLanguageId,
		initialDataType,
		validation,
		validations,
		value,
	});

	return (
		<Validation
			{...data}
			defaultLanguageId={defaultLanguageId}
			editingLanguageId={editingLanguageId}
			label={label}
			name={name}
			onBlur={onBlur}
			onChange={(value) => onChange({target: {value}})}
			readOnly={readOnly}
			validation={validation}
			value={value}
			visible={visible}
		/>
	);
};

export default ValidationWrapper;
