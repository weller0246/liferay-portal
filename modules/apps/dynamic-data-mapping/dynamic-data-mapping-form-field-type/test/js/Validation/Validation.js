/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import {cleanup, render} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import {FormProvider} from 'data-engine-js-components-web';
import React from 'react';

import Validation from '../../../src/main/resources/META-INF/resources/Validation/Validation';

const globalLanguageDirection = Liferay.Language.direction;

const spritemap = 'icons.svg';

const defaultValue = {
	errorMessage: {},
	expression: {},
	parameter: {},
};

const ValidationWithProvider = ({validations, ...props}) => (
	<FormProvider initialState={{validations}}>
		<Validation {...props} />
	</FormProvider>
);

describe('Validation', () => {
	beforeAll(() => {
		Liferay.Language.direction = {
			en_US: 'rtl',
		};
	});

	afterAll(() => {
		Liferay.Language.direction = globalLanguageDirection;
	});

	afterEach(cleanup);

	it('renders checkbox to enable Validation', () => {
		const {container} = render(
			<ValidationWithProvider
				dataType="string"
				label="Validator"
				name="validation"
				onChange={() => {}}
				spritemap={spritemap}
				validations={{
					string: [
						{
							label: '',
							name: '',
							parameterMessage: '',
							template: '',
						},
					],
				}}
				value={defaultValue}
			/>
		);

		expect(container).toMatchSnapshot();
	});

	it('enables validation after click on toogle', () => {
		const onChange = jest.fn();

		const {container} = render(
			<ValidationWithProvider
				dataType="string"
				defaultLanguageId="en_US"
				editingLanguageId="en_US"
				expression={{}}
				label="Validator"
				name="validation"
				onChange={onChange}
				spritemap={spritemap}
				validation={{
					dataType: 'string',
					fieldName: 'textfield',
				}}
				validations={{
					string: [
						{
							label: '',
							name: 'contains',
							parameterMessage: '',
							template: 'contains({name}, "{parameter}")',
						},
					],
				}}
				value={defaultValue}
			/>
		);

		const inputCheckbox = container.querySelector('input[type="checkbox"]');

		userEvent.click(inputCheckbox);

		expect(onChange).toHaveBeenLastCalledWith({
			target: {
				value: {
					enableValidation: true,
					errorMessage: {
						en_US: undefined,
					},
					expression: {
						name: 'contains',
						value: 'contains(textfield, "{parameter}")',
					},
					parameter: {
						en_US: undefined,
					},
				},
			},
		});
	});

	it('renders parameter field with Numeric element', () => {
		const onChange = jest.fn();

		const {container} = render(
			<ValidationWithProvider
				dataType="numeric"
				defaultLanguageId="en_US"
				editingLanguageId="en_US"
				expression={{}}
				label="Validator"
				name="validation"
				onChange={onChange}
				spritemap={spritemap}
				validation={{
					dataType: 'integer',
					fieldName: 'numericfield',
				}}
				validations={{
					numeric: [
						{
							label: '',
							name: 'eq',
							parameterMessage: '',
							template: '{name}=={parameter}',
						},
					],
				}}
				value={defaultValue}
			/>
		);

		const inputCheckbox = container.querySelector('input[type="checkbox"]');

		userEvent.click(inputCheckbox);

		expect(onChange).toHaveBeenLastCalledWith({
			target: {
				value: {
					enableValidation: true,
					errorMessage: {
						en_US: undefined,
					},
					expression: {
						name: 'eq',
						value: 'numericfield=={parameter}',
					},
					parameter: {
						en_US: undefined,
					},
				},
			},
		});
	});

	it('renders parameter field with Date element', () => {
		const onChange = jest.fn();

		const {container} = render(
			<ValidationWithProvider
				dataType="date"
				defaultLanguageId="en_US"
				editingLanguageId="en_US"
				expression={{}}
				label="Validator"
				name="validation"
				onChange={onChange}
				spritemap={spritemap}
				validation={{
					dataType: 'date',
					fieldName: 'dateField',
				}}
				validations={{
					date: [
						{
							label: '',
							name: 'futureDates',
							parameterMessage: '',
							template: 'futureDates({name}, "{parameter}")',
						},
					],
				}}
				value={defaultValue}
			/>
		);

		const inputCheckbox = container.querySelector('input[type="checkbox"]');

		userEvent.click(inputCheckbox);

		expect(onChange).toHaveBeenLastCalledWith({
			target: {
				value: {
					enableValidation: true,
					errorMessage: {
						en_US: undefined,
					},
					expression: {
						name: 'futureDates',
						value: 'futureDates(dateField, "{parameter}")',
					},
					parameter: {
						en_US: undefined,
					},
				},
			},
		});
	});
});
