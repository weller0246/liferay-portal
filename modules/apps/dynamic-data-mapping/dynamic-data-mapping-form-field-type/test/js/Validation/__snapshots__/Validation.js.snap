// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validation renders checkbox to enable Validation 1`] = `
<div>
  <div
    class="form-group lfr-ddm-form-field-validation"
  >
    <div
      aria-labelledby="enableValidation"
      class="form-group hide"
      data-field-name="enableValidation"
      tabindex="0"
    >
      <label
        class="ddm-toggle-switch toggle-switch"
      >
        <input
          class="toggle-switch-check"
          name="enableValidation"
          type="checkbox"
          value="true"
        />
        <span
          aria-hidden="true"
          class="toggle-switch-bar"
        >
          <span
            class="toggle-switch-handle"
          />
          <span
            class="toggle-switch-text toggle-switch-text-right"
          >
            Validator
          </span>
        </span>
      </label>
      <input
        name="enableValidation_edited"
        type="hidden"
        value="false"
      />
      
    </div>
  </div>
</div>
`;
